{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3e008d01-0c16-41d3-a57b-99b506be5223",
   "metadata": {},
   "source": [
    "Q1. Explain why we have to use the Exception class while creating a Custom Exception.\n",
    "\n",
    "In Python, custom exceptions are created by defining new classes that inherit from the built-in Exception class (or its subclasses). This is a good practice for several reasons:\n",
    "\n",
    "Clarity and Readability: Custom exceptions help make your code more readable and maintainable. When an exception occurs, it's easier to understand what went wrong if the exception class name is descriptive and meaningful.\n",
    "\n",
    "Granular Error Handling: Custom exceptions allow you to handle specific error scenarios in your code. By defining distinct exception classes, you can catch and handle different types of errors separately, providing more precise error messages or actions.\n",
    "\n",
    "Code Organization: Grouping related exceptions into custom exception hierarchies helps in organizing your code and its error handling logic. This makes it easier to locate and manage error-handling code.\n",
    "\n",
    "Debugging: Custom exceptions can provide additional context or information about the error, which can aid in debugging. You can include attributes or methods in your custom exceptions to carry relevant data or messages.\n",
    "\n",
    "Q2. Write a python program to print Python Exception Hierarchy.\n",
    "\n",
    "Here's a simple Python program that prints the exception hierarchy:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "09e0e783-bf6f-4f5f-97ad-5d28d6b83ecb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'BaseException'>\n",
      "    <class 'Exception'>\n",
      "        <class 'TypeError'>\n",
      "            <class 'decimal.FloatOperation'>\n",
      "            <class 'email.errors.MultipartConversionError'>\n",
      "        <class 'StopAsyncIteration'>\n",
      "        <class 'StopIteration'>\n",
      "        <class 'ImportError'>\n",
      "            <class 'ModuleNotFoundError'>\n",
      "            <class 'zipimport.ZipImportError'>\n",
      "        <class 'OSError'>\n",
      "            <class 'ConnectionError'>\n",
      "                <class 'BrokenPipeError'>\n",
      "                <class 'ConnectionAbortedError'>\n",
      "                <class 'ConnectionRefusedError'>\n",
      "                <class 'ConnectionResetError'>\n",
      "                    <class 'http.client.RemoteDisconnected'>\n",
      "            <class 'BlockingIOError'>\n",
      "            <class 'ChildProcessError'>\n",
      "            <class 'FileExistsError'>\n",
      "            <class 'FileNotFoundError'>\n",
      "            <class 'IsADirectoryError'>\n",
      "            <class 'NotADirectoryError'>\n",
      "            <class 'InterruptedError'>\n",
      "                <class 'zmq.error.InterruptedSystemCall'>\n",
      "            <class 'PermissionError'>\n",
      "            <class 'ProcessLookupError'>\n",
      "            <class 'TimeoutError'>\n",
      "            <class 'io.UnsupportedOperation'>\n",
      "            <class 'signal.itimer_error'>\n",
      "            <class 'socket.herror'>\n",
      "            <class 'socket.gaierror'>\n",
      "            <class 'ssl.SSLError'>\n",
      "                <class 'ssl.SSLCertVerificationError'>\n",
      "                <class 'ssl.SSLZeroReturnError'>\n",
      "                <class 'ssl.SSLWantWriteError'>\n",
      "                <class 'ssl.SSLWantReadError'>\n",
      "                <class 'ssl.SSLSyscallError'>\n",
      "                <class 'ssl.SSLEOFError'>\n",
      "            <class 'shutil.Error'>\n",
      "                <class 'shutil.SameFileError'>\n",
      "            <class 'shutil.SpecialFileError'>\n",
      "            <class 'shutil.ExecError'>\n",
      "            <class 'shutil.ReadError'>\n",
      "            <class 'urllib.error.URLError'>\n",
      "                <class 'urllib.error.HTTPError'>\n",
      "                <class 'urllib.error.ContentTooShortError'>\n",
      "            <class 'gzip.BadGzipFile'>\n",
      "        <class 'EOFError'>\n",
      "            <class 'asyncio.exceptions.IncompleteReadError'>\n",
      "        <class 'RuntimeError'>\n",
      "            <class 'RecursionError'>\n",
      "            <class 'NotImplementedError'>\n",
      "                <class 'zmq.error.ZMQVersionError'>\n",
      "                <class 'IPython.core.error.StdinNotImplementedError'>\n",
      "            <class '_frozen_importlib._DeadlockError'>\n",
      "            <class 'threading.BrokenBarrierError'>\n",
      "            <class 'concurrent.futures._base.BrokenExecutor'>\n",
      "                <class 'concurrent.futures.thread.BrokenThreadPool'>\n",
      "            <class 'asyncio.exceptions.SendfileNotAvailableError'>\n",
      "            <class 'pkg_resources.ExtractionError'>\n",
      "            <class '_pydevd_bundle.pydevd_vars.VariableError'>\n",
      "        <class 'NameError'>\n",
      "            <class 'UnboundLocalError'>\n",
      "        <class 'AttributeError'>\n",
      "            <class 'dataclasses.FrozenInstanceError'>\n",
      "        <class 'SyntaxError'>\n",
      "            <class 'IndentationError'>\n",
      "                <class 'TabError'>\n",
      "        <class 'LookupError'>\n",
      "            <class 'IndexError'>\n",
      "            <class 'KeyError'>\n",
      "                <class 'jupyter_client.kernelspec.NoSuchKernel'>\n",
      "                <class 'IPython.terminal.pt_inputhooks.UnknownBackend'>\n",
      "            <class 'encodings.CodecRegistryError'>\n",
      "        <class 'ValueError'>\n",
      "            <class 'UnicodeError'>\n",
      "                <class 'UnicodeEncodeError'>\n",
      "                <class 'UnicodeDecodeError'>\n",
      "                <class 'UnicodeTranslateError'>\n",
      "            <class 'io.UnsupportedOperation'>\n",
      "            <class 'json.decoder.JSONDecodeError'>\n",
      "            <class 'ssl.SSLCertVerificationError'>\n",
      "            <class 'binascii.Error'>\n",
      "            <class '_hashlib.UnsupportedDigestmodError'>\n",
      "            <class 'calendar.IllegalMonthError'>\n",
      "            <class 'calendar.IllegalWeekdayError'>\n",
      "            <class 'dateutil.parser._parser.ParserError'>\n",
      "            <class 'pygments.util.ClassNotFound'>\n",
      "            <class 'IPython.lib.clipboard.ClipboardEmpty'>\n",
      "            <class 'email.errors.MessageDefect'>\n",
      "                <class 'email.errors.NoBoundaryInMultipartDefect'>\n",
      "                <class 'email.errors.StartBoundaryNotFoundDefect'>\n",
      "                <class 'email.errors.CloseBoundaryNotFoundDefect'>\n",
      "                <class 'email.errors.FirstHeaderLineIsContinuationDefect'>\n",
      "                <class 'email.errors.MisplacedEnvelopeHeaderDefect'>\n",
      "                <class 'email.errors.MissingHeaderBodySeparatorDefect'>\n",
      "                <class 'email.errors.MultipartInvariantViolationDefect'>\n",
      "                <class 'email.errors.InvalidMultipartContentTransferEncodingDefect'>\n",
      "                <class 'email.errors.UndecodableBytesDefect'>\n",
      "                <class 'email.errors.InvalidBase64PaddingDefect'>\n",
      "                <class 'email.errors.InvalidBase64CharactersDefect'>\n",
      "                <class 'email.errors.InvalidBase64LengthDefect'>\n",
      "                <class 'email.errors.HeaderDefect'>\n",
      "                    <class 'email.errors.InvalidHeaderDefect'>\n",
      "                    <class 'email.errors.HeaderMissingRequiredValue'>\n",
      "                    <class 'email.errors.NonPrintableDefect'>\n",
      "                    <class 'email.errors.ObsoleteHeaderDefect'>\n",
      "                    <class 'email.errors.NonASCIILocalPartDefect'>\n",
      "                    <class 'email.errors.InvalidDateDefect'>\n",
      "            <class 'IPython.core.magics.code.MacroToEdit'>\n",
      "            <class 'plistlib.InvalidFileException'>\n",
      "            <class 'pkg_resources._vendor.more_itertools.more.UnequalIterablesError'>\n",
      "            <class 'pkg_resources.extern.packaging.version.InvalidVersion'>\n",
      "            <class 'pkg_resources._vendor.packaging._manylinux._ELFFileHeader._InvalidELFFileHeader'>\n",
      "            <class 'pkg_resources.extern.packaging.utils.InvalidWheelFilename'>\n",
      "            <class 'pkg_resources.extern.packaging.utils.InvalidSdistFilename'>\n",
      "            <class 'pkg_resources.extern.packaging.specifiers.InvalidSpecifier'>\n",
      "            <class 'pkg_resources.extern.packaging.markers.InvalidMarker'>\n",
      "            <class 'pkg_resources.extern.packaging.markers.UndefinedComparison'>\n",
      "            <class 'pkg_resources.extern.packaging.markers.UndefinedEnvironmentName'>\n",
      "            <class 'pkg_resources.extern.packaging.requirements.InvalidRequirement'>\n",
      "                <class 'pkg_resources.RequirementParseError'>\n",
      "            <class 'packaging.version.InvalidVersion'>\n",
      "        <class 'AssertionError'>\n",
      "        <class 'ArithmeticError'>\n",
      "            <class 'FloatingPointError'>\n",
      "            <class 'OverflowError'>\n",
      "            <class 'ZeroDivisionError'>\n",
      "                <class 'decimal.DivisionByZero'>\n",
      "                <class 'decimal.DivisionUndefined'>\n",
      "            <class 'decimal.DecimalException'>\n",
      "                <class 'decimal.Clamped'>\n",
      "                <class 'decimal.Rounded'>\n",
      "                    <class 'decimal.Underflow'>\n",
      "                    <class 'decimal.Overflow'>\n",
      "                <class 'decimal.Inexact'>\n",
      "                    <class 'decimal.Underflow'>\n",
      "                    <class 'decimal.Overflow'>\n",
      "                <class 'decimal.Subnormal'>\n",
      "                    <class 'decimal.Underflow'>\n",
      "                <class 'decimal.DivisionByZero'>\n",
      "                <class 'decimal.FloatOperation'>\n",
      "                <class 'decimal.InvalidOperation'>\n",
      "                    <class 'decimal.ConversionSyntax'>\n",
      "                    <class 'decimal.DivisionImpossible'>\n",
      "                    <class 'decimal.DivisionUndefined'>\n",
      "                    <class 'decimal.InvalidContext'>\n",
      "        <class 'SystemError'>\n",
      "            <class 'encodings.CodecRegistryError'>\n",
      "        <class 'ReferenceError'>\n",
      "        <class 'MemoryError'>\n",
      "        <class 'BufferError'>\n",
      "        <class 'Warning'>\n",
      "            <class 'UserWarning'>\n",
      "                <class 'getpass.GetPassWarning'>\n",
      "                <class 'IPython.core.formatters.FormatterWarning'>\n",
      "            <class 'EncodingWarning'>\n",
      "            <class 'DeprecationWarning'>\n",
      "                <class 'IPython.core.interactiveshell.ProvisionalWarning'>\n",
      "            <class 'PendingDeprecationWarning'>\n",
      "            <class 'SyntaxWarning'>\n",
      "            <class 'RuntimeWarning'>\n",
      "                <class 'zmq.asyncio.ProactorSelectorThreadWarning'>\n",
      "                <class 'dateutil.parser._parser.UnknownTimezoneWarning'>\n",
      "                <class 'pkg_resources.PEP440Warning'>\n",
      "            <class 'FutureWarning'>\n",
      "                <class 'IPython.core.completer.ProvisionalCompleterWarning'>\n",
      "            <class 'ImportWarning'>\n",
      "            <class 'UnicodeWarning'>\n",
      "            <class 'BytesWarning'>\n",
      "            <class 'ResourceWarning'>\n",
      "            <class 'dateutil.tz.DeprecatedTzFormatWarning'>\n",
      "            <class 'pkg_resources.PkgResourcesDeprecationWarning'>\n",
      "        <class 'warnings._OptionError'>\n",
      "        <class 'runpy._Error'>\n",
      "        <class 're.error'>\n",
      "        <class 'sre_parse.Verbose'>\n",
      "        <class 'locale.Error'>\n",
      "        <class 'subprocess.SubprocessError'>\n",
      "            <class 'subprocess.CalledProcessError'>\n",
      "            <class 'subprocess.TimeoutExpired'>\n",
      "        <class 'tokenize.TokenError'>\n",
      "        <class 'tokenize.StopTokenizing'>\n",
      "        <class 'inspect.ClassFoundException'>\n",
      "        <class 'inspect.EndOfBlock'>\n",
      "        <class 'traitlets.traitlets.TraitError'>\n",
      "        <class 'copy.Error'>\n",
      "        <class 'concurrent.futures._base.Error'>\n",
      "            <class 'concurrent.futures._base.CancelledError'>\n",
      "            <class 'concurrent.futures._base.TimeoutError'>\n",
      "            <class 'concurrent.futures._base.InvalidStateError'>\n",
      "        <class 'socket._GiveupOnSendfile'>\n",
      "        <class 'struct.error'>\n",
      "        <class 'binascii.Incomplete'>\n",
      "        <class 'asyncio.exceptions.TimeoutError'>\n",
      "        <class 'asyncio.exceptions.InvalidStateError'>\n",
      "        <class 'asyncio.exceptions.LimitOverrunError'>\n",
      "        <class 'asyncio.queues.QueueEmpty'>\n",
      "        <class 'asyncio.queues.QueueFull'>\n",
      "        <class '_queue.Empty'>\n",
      "        <class 'queue.Full'>\n",
      "        <class 'ctypes.ArgumentError'>\n",
      "        <class 'zmq.error.ZMQBaseError'>\n",
      "            <class 'zmq.error.ZMQError'>\n",
      "                <class 'zmq.error.ContextTerminated'>\n",
      "                <class 'zmq.error.Again'>\n",
      "                <class 'zmq.error.InterruptedSystemCall'>\n",
      "            <class 'zmq.error.ZMQBindError'>\n",
      "            <class 'zmq.error.NotDone'>\n",
      "        <class '_pickle.PickleError'>\n",
      "            <class '_pickle.PicklingError'>\n",
      "            <class '_pickle.UnpicklingError'>\n",
      "        <class 'pickle._Stop'>\n",
      "        <class 'argparse.ArgumentError'>\n",
      "        <class 'argparse.ArgumentTypeError'>\n",
      "        <class 'traitlets.config.loader.ConfigError'>\n",
      "            <class 'traitlets.config.loader.ConfigLoaderError'>\n",
      "                <class 'traitlets.config.loader.ArgumentError'>\n",
      "            <class 'traitlets.config.loader.ConfigFileNotFound'>\n",
      "        <class 'traitlets.config.configurable.ConfigurableError'>\n",
      "            <class 'traitlets.config.configurable.MultipleInstanceError'>\n",
      "        <class 'traitlets.config.application.ApplicationError'>\n",
      "        <class 'zlib.error'>\n",
      "        <class 'tornado.util.TimeoutError'>\n",
      "        <class '_curses.error'>\n",
      "        <class 'tornado.concurrent.ReturnValueIgnoredError'>\n",
      "        <class 'tornado.gen.KeyReuseError'>\n",
      "        <class 'tornado.gen.UnknownKeyError'>\n",
      "        <class 'tornado.gen.LeakedCallbackError'>\n",
      "        <class 'tornado.gen.BadYieldError'>\n",
      "        <class 'tornado.gen.ReturnValueIgnoredError'>\n",
      "        <class 'tornado.gen.Return'>\n",
      "        <class 'jupyter_client.channels.InvalidPortNumber'>\n",
      "        <class 'termios.error'>\n",
      "        <class '_lzma.LZMAError'>\n",
      "        <class 'shutil.RegistryError'>\n",
      "        <class 'shutil._GiveupOnFastCopy'>\n",
      "        <class 'configparser.Error'>\n",
      "            <class 'configparser.NoSectionError'>\n",
      "            <class 'configparser.DuplicateSectionError'>\n",
      "            <class 'configparser.DuplicateOptionError'>\n",
      "            <class 'configparser.NoOptionError'>\n",
      "            <class 'configparser.InterpolationError'>\n",
      "                <class 'configparser.InterpolationMissingOptionError'>\n",
      "                <class 'configparser.InterpolationSyntaxError'>\n",
      "                <class 'configparser.InterpolationDepthError'>\n",
      "            <class 'configparser.ParsingError'>\n",
      "                <class 'configparser.MissingSectionHeaderError'>\n",
      "        <class 'jupyter_client.localinterfaces.NoIPAddresses'>\n",
      "        <class 'zipfile.BadZipFile'>\n",
      "        <class 'zipfile.LargeZipFile'>\n",
      "        <class 'entrypoints.BadEntryPoint'>\n",
      "        <class 'entrypoints.NoSuchEntryPoint'>\n",
      "        <class 'jupyter_client.multikernelmanager.DuplicateKernelError'>\n",
      "        <class 'pydoc.ErrorDuringImport'>\n",
      "        <class 'executing.executing.NotOneValueFound'>\n",
      "        <class 'pure_eval.utils.CannotEval'>\n",
      "        <class 'pygments.util.OptionError'>\n",
      "        <class 'bdb.BdbQuit'>\n",
      "        <class 'pdb.Restart'>\n",
      "        <class 'pexpect.exceptions.ExceptionPexpect'>\n",
      "            <class 'pexpect.exceptions.EOF'>\n",
      "            <class 'pexpect.exceptions.TIMEOUT'>\n",
      "        <class 'ptyprocess.util.PtyProcessError'>\n",
      "        <class 'IPython.utils.process.FindCmdError'>\n",
      "        <class 'IPython.utils.path.HomeDirError'>\n",
      "        <class 'IPython.core.profiledir.ProfileDirError'>\n",
      "        <class 'IPython.core.error.IPythonCoreError'>\n",
      "            <class 'IPython.core.error.TryNext'>\n",
      "            <class 'IPython.core.error.UsageError'>\n",
      "            <class 'IPython.core.error.StdinNotImplementedError'>\n",
      "        <class 'IPython.core.error.InputRejected'>\n",
      "        <class 'getopt.GetoptError'>\n",
      "        <class 'pygments.filters.ErrorToken'>\n",
      "        <class 'IPython.core.prefilter.PrefilterError'>\n",
      "        <class 'IPython.core.alias.AliasError'>\n",
      "            <class 'IPython.core.alias.InvalidAliasError'>\n",
      "        <class 'sqlite3.Error'>\n",
      "            <class 'sqlite3.InterfaceError'>\n",
      "            <class 'sqlite3.DatabaseError'>\n",
      "                <class 'sqlite3.InternalError'>\n",
      "                <class 'sqlite3.OperationalError'>\n",
      "                <class 'sqlite3.ProgrammingError'>\n",
      "                <class 'sqlite3.IntegrityError'>\n",
      "                <class 'sqlite3.DataError'>\n",
      "                <class 'sqlite3.NotSupportedError'>\n",
      "        <class 'sqlite3.Warning'>\n",
      "        <class 'IPython.core.interactiveshell.SpaceInInput'>\n",
      "        <class 'xml.dom.DOMException'>\n",
      "            <class 'xml.dom.IndexSizeErr'>\n",
      "            <class 'xml.dom.DomstringSizeErr'>\n",
      "            <class 'xml.dom.HierarchyRequestErr'>\n",
      "            <class 'xml.dom.WrongDocumentErr'>\n",
      "            <class 'xml.dom.InvalidCharacterErr'>\n",
      "            <class 'xml.dom.NoDataAllowedErr'>\n",
      "            <class 'xml.dom.NoModificationAllowedErr'>\n",
      "            <class 'xml.dom.NotFoundErr'>\n",
      "            <class 'xml.dom.NotSupportedErr'>\n",
      "            <class 'xml.dom.InuseAttributeErr'>\n",
      "            <class 'xml.dom.InvalidStateErr'>\n",
      "            <class 'xml.dom.SyntaxErr'>\n",
      "            <class 'xml.dom.InvalidModificationErr'>\n",
      "            <class 'xml.dom.NamespaceErr'>\n",
      "            <class 'xml.dom.InvalidAccessErr'>\n",
      "            <class 'xml.dom.ValidationErr'>\n",
      "        <class 'prompt_toolkit.validation.ValidationError'>\n",
      "        <class 'prompt_toolkit.buffer.EditReadOnlyBuffer'>\n",
      "        <class 'prompt_toolkit.buffer._Retry'>\n",
      "        <class 'prompt_toolkit.layout.layout.InvalidLayoutError'>\n",
      "        <class 'prompt_toolkit.renderer.HeightIsUnknownError'>\n",
      "        <class 'parso.parser.ParserSyntaxError'>\n",
      "        <class 'parso.parser.InternalParseError'>\n",
      "        <class 'parso.python.diff._PositionUpdatingFinished'>\n",
      "        <class 'jedi.inference.helpers.SimpleGetItemNotFound'>\n",
      "        <class 'jedi.inference.utils.UncaughtAttributeError'>\n",
      "        <class 'jedi.inference.base_value.HasNoContext'>\n",
      "        <class 'jedi.inference.arguments.ParamIssue'>\n",
      "        <class 'jedi.api.exceptions._JediError'>\n",
      "            <class 'jedi.api.exceptions.InternalError'>\n",
      "            <class 'jedi.api.exceptions.WrongVersion'>\n",
      "            <class 'jedi.api.exceptions.RefactoringError'>\n",
      "        <class 'jedi.api.helpers.OnErrorLeaf'>\n",
      "        <class 'jedi.api.environment.InvalidPythonEnvironment'>\n",
      "        <class 'email.errors.MessageError'>\n",
      "            <class 'email.errors.MessageParseError'>\n",
      "                <class 'email.errors.HeaderParseError'>\n",
      "                <class 'email.errors.BoundaryError'>\n",
      "            <class 'email.errors.MultipartConversionError'>\n",
      "            <class 'email.errors.CharsetError'>\n",
      "        <class 'uu.Error'>\n",
      "        <class 'http.client.HTTPException'>\n",
      "            <class 'http.client.NotConnected'>\n",
      "            <class 'http.client.InvalidURL'>\n",
      "            <class 'http.client.UnknownProtocol'>\n",
      "            <class 'http.client.UnknownTransferEncoding'>\n",
      "            <class 'http.client.UnimplementedFileMode'>\n",
      "            <class 'http.client.IncompleteRead'>\n",
      "            <class 'http.client.ImproperConnectionState'>\n",
      "                <class 'http.client.CannotSendRequest'>\n",
      "                <class 'http.client.CannotSendHeader'>\n",
      "                <class 'http.client.ResponseNotReady'>\n",
      "            <class 'http.client.BadStatusLine'>\n",
      "                <class 'http.client.RemoteDisconnected'>\n",
      "            <class 'http.client.LineTooLong'>\n",
      "        <class 'IPython.core.magics.code.InteractivelyDefined'>\n",
      "        <class 'IPython.terminal.embed.KillEmbedded'>\n",
      "        <class 'psutil.Error'>\n",
      "            <class 'psutil.NoSuchProcess'>\n",
      "                <class 'psutil.ZombieProcess'>\n",
      "            <class 'psutil.AccessDenied'>\n",
      "            <class 'psutil.TimeoutExpired'>\n",
      "        <class 'psutil._pslinux._Ipv6UnsupportedError'>\n",
      "        <class 'tornado.queues.QueueEmpty'>\n",
      "        <class 'tornado.queues.QueueFull'>\n",
      "        <class '_pydev_bundle._pydev_saved_modules.DebuggerInitializationError'>\n",
      "        <class 'xml.parsers.expat.ExpatError'>\n",
      "        <class 'xmlrpc.client.Error'>\n",
      "            <class 'xmlrpc.client.ProtocolError'>\n",
      "            <class 'xmlrpc.client.ResponseError'>\n",
      "            <class 'xmlrpc.client.Fault'>\n",
      "        <class 'pkg_resources._vendor.pyparsing.exceptions.ParseBaseException'>\n",
      "            <class 'pkg_resources._vendor.pyparsing.exceptions.ParseException'>\n",
      "            <class 'pkg_resources._vendor.pyparsing.exceptions.ParseFatalException'>\n",
      "                <class 'pkg_resources._vendor.pyparsing.exceptions.ParseSyntaxException'>\n",
      "        <class 'pkg_resources._vendor.pyparsing.exceptions.RecursiveGrammarException'>\n",
      "        <class 'pkg_resources.ResolutionError'>\n",
      "            <class 'pkg_resources.VersionConflict'>\n",
      "                <class 'pkg_resources.ContextualVersionConflict'>\n",
      "            <class 'pkg_resources.DistributionNotFound'>\n",
      "            <class 'pkg_resources.UnknownExtra'>\n",
      "        <class '_pydevd_bundle.pydevd_runpy._Error'>\n",
      "        <class '_pydevd_bundle.pydevd_resolver.UnableToResolveVariableException'>\n",
      "        <class '_pydev_bundle.pydev_monkey.InvalidTypeInArgsException'>\n",
      "    <class 'GeneratorExit'>\n",
      "    <class 'SystemExit'>\n",
      "    <class 'KeyboardInterrupt'>\n",
      "    <class 'asyncio.exceptions.CancelledError'>\n",
      "    <class 'pkg_resources._vendor.more_itertools.more.AbortThread'>\n"
     ]
    }
   ],
   "source": [
    "def print_exception_hierarchy(exception_class, depth=0):\n",
    "    indent = ' ' * (depth * 4)\n",
    "    print(indent + str(exception_class))\n",
    "    \n",
    "    for subclass in exception_class.__subclasses__():\n",
    "        print_exception_hierarchy(subclass, depth + 1)\n",
    "\n",
    "print_exception_hierarchy(BaseException)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9325dea-a109-4fdd-80e4-75e79af1ff59",
   "metadata": {},
   "source": [
    "Q3. What errors are defined in the ArithmeticError class? Explain any two with an example.\n",
    "\n",
    "The ArithmeticError class is a base class for arithmetic-related exceptions in Python. Two common exceptions derived from ArithmeticError are ZeroDivisionError and OverflowError.\n",
    "\n",
    "ZeroDivisionError:\n",
    "Raised when attempting to divide by zero."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "388f7741-b2f4-4135-ba59-4e8fead1535d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: division by zero\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    result = 10 / 0\n",
    "except ZeroDivisionError as e:\n",
    "    print(\"Error:\", e)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18dc744b-c403-497c-ac24-1678e863d728",
   "metadata": {},
   "source": [
    "OverflowError:\n",
    "\n",
    "Raised when an arithmetic operation exceeds the limits of the data type."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f114defb-1d23-487f-aa03-a2eecdd1e4db",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "try:\n",
    "    large_number = sys.maxsize + 1\n",
    "except OverflowError as e:\n",
    "    print(\"Error:\", e)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b85b5bc-9380-4d43-95bd-c79ab50702ea",
   "metadata": {},
   "source": [
    "Q4. Why LookupError class is used? Explain with an example KeyError and IndexError.\n",
    "\n",
    "The LookupError class is a base class for exceptions that occur when an index or key is not found in a sequence or mapping. It encompasses errors like KeyError and IndexError.\n",
    "\n",
    "KeyError:\n",
    "Raised when a dictionary key is not found."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "346a7ec1-766e-4aaa-9fc3-4cf97f8caeb8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: 'c'\n"
     ]
    }
   ],
   "source": [
    "my_dict = {'a': 1, 'b': 2}\n",
    "try:\n",
    "    value = my_dict['c']\n",
    "except KeyError as e:\n",
    "    print(\"Error:\", e)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73a75c3d-7e0c-4bf5-9147-57305f3bc879",
   "metadata": {},
   "source": [
    "IndexError:\n",
    "    \n",
    "Raised when an index is out of range in a sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dd737951-f2bc-4127-b88f-7625cb84e07e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: list index out of range\n"
     ]
    }
   ],
   "source": [
    "my_list = [1, 2, 3]\n",
    "try:\n",
    "    value = my_list[5]\n",
    "except IndexError as e:\n",
    "    print(\"Error:\", e)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31ef9957-1075-4392-bee0-b7382c8b7af6",
   "metadata": {},
   "source": [
    "Q5. Explain ImportError. What is ModuleNotFoundError?\n",
    "\n",
    "ImportError is a base class for exceptions raised when an import statement fails. It can occur if the specified module or attribute cannot be imported.\n",
    "\n",
    "ModuleNotFoundError is a specific subclass of ImportError that is raised when a module could not be found during import. It was introduced in Python 3.6 to provide more accurate information about missing modules.\n",
    "\n",
    "Q6. List down some best practices for exception handling in Python.\n",
    "\n",
    "Here are some best practices for exception handling in Python:\n",
    "\n",
    "Specific Exception Handling: Catch specific exceptions instead of using broad exceptions like Exception or BaseException. This allows you to handle errors more precisely.\n",
    "\n",
    "Use try-except Blocks: Wrap potentially error-prone code in try-except blocks. This helps isolate and handle exceptions gracefully.\n",
    "\n",
    "Avoid Bare except: Avoid using bare except clauses (without specifying the exception type). This can hide unexpected errors and make debugging difficult.\n",
    "\n",
    "Cleanup with finally: Use the finally block to ensure cleanup code (e.g., closing files) is executed regardless of whether an exception occurred.\n",
    "\n",
    "Raising Exceptions: Raise exceptions that convey meaningful information about the error. Include relevant error messages or context.\n",
    "\n",
    "Avoid Deep Nesting: Avoid excessive nesting of try-except blocks. It can make code harder to read and understand.\n",
    "\n",
    "Logging: Utilize Python's logging module to record error information. This helps in debugging and monitoring.\n",
    "\n",
    "Custom Exception Classes: Use custom exception classes for specific errors to provide better context and clarity.\n",
    "\n",
    "Don't Suppress Errors: Avoid catching exceptions and ignoring them without proper handling. If you can't handle an exception properly, consider letting it propagate.\n",
    "\n",
    "Keep It Simple: Keep your exception handling logic straightforward. Complex exception handling can make your code harder to maintain.\n",
    "\n",
    "Remember, the goal of exception handling is to make your code robust and user-friendly by gracefully managing unexpected situations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f6003a9-2ee5-4a8a-aa98-800e23b5942d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
